app/components/autocomplete/channel_mention/channel_mention.js

```diff
@@ -0,0 +1,266 @@
+
+import {RequestStatus} from 'service/constants';
+
+const CHANNEL_MENTION_REGEX = /\B(~([^~\r\n]*))$/i;
+
+const getStyleFromTheme = makeStyleSheetFromTheme((theme) => {
+    return StyleSheet.create({
+        section: {
+            justifyContent: 'center',
+            paddingLeft: 8,
+            backgroundColor: changeOpacity(theme.centerChannelColor, 0.1),
+            borderTopWidth: 1,
+            borderTopColor: changeOpacity(theme.centerChannelColor, 0.2)
+        },
+        sectionText: {
+            fontSize: 12,
+            color: changeOpacity(theme.centerChannelColor, 0.7),
+            paddingVertical: 7
+        },
+        sectionWrapper: {
+            backgroundColor: theme.centerChannelBg
+        },
+        listView: {
+            flex: 1,
+            backgroundColor: theme.centerChannelBg,
+            borderWidth: 1,
+            borderColor: changeOpacity(theme.centerChannelColor, 0.2),
+            borderTopWidth: 0,
+            borderBottomWidth: 0
+        },
+        loading: {
+            alignItems: 'center',
+            justifyContent: 'center',
+            paddingVertical: 20,
+            backgroundColor: theme.centerChannelBg,
+            borderWidth: 1,
+            borderColor: changeOpacity(theme.centerChannelColor, 0.2),
+            borderBottomWidth: 0
+        },
+        row: {
+            padding: 8,
+            flexDirection: 'row',
+            alignItems: 'center',
+            backgroundColor: theme.centerChannelBg,
+            borderTopWidth: 1,
+            borderTopColor: changeOpacity(theme.centerChannelColor, 0.2)
+        },
+        rowDisplayName: {
+            fontSize: 13,
+            color: theme.centerChannelColor
+        },
+        rowName: {
+            color: theme.centerChannelColor,
+            opacity: 0.6
+        }
+    });
+});
+
+export default class ChannelMention extends Component {
+    static propTypes = {
+        currentChannelId: PropTypes.string.isRequired,
+        currentTeamId: PropTypes.string.isRequired,
+        cursorPosition: PropTypes.number.isRequired,
+        autocompleteChannels: PropTypes.object.isRequired,
+        postDraft: PropTypes.string,
+        requestStatus: PropTypes.string.isRequired,
+        theme: PropTypes.object.isRequired,
+        actions: PropTypes.shape({
+            changePostDraft: PropTypes.func.isRequired,
+            autocompleteChannels: PropTypes.func.isRequired
+        })
+    }
+
+    static defaultProps = {
+        postDraft: ''
+    }
+
+    constructor(props) {
+        super(props);
+
+        const ds = new ListView.DataSource({
+            sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
+            rowHasChanged: (r1, r2) => r1 !== r2
+        });
+
+        this.state = {
+            active: false,
+            dataSource: ds.cloneWithRowsAndSections(props.autocompleteChannels)
+        };
+    }
+
+    componentWillReceiveProps(nextProps) {
```

Comment:
I just tried placing console.logs() throughout ```componentWillReceiveProps``` and the only time that they showed up was when we had a match based on the regex even when I would perform other actions to change the store so I think that it is dependent only on the postDraft changing since we return if we don't have a match. Are you seeing something differently on your end?